README.md
# rostate_machine
state machine library for ROS

![Developed By OUXT Polaris](img/logo.png "Logo")

| *master* | *develop* |
|----------|-----------|
|[![Build Status](https://travis-ci.org/OUXT-Polaris/rostate_machine.svg?branch=master)](https://travis-ci.org/OUXT-Polaris/rostate_machine)|[![Build Status](https://travis-ci.org/OUXT-Polaris/rostate_machine.svg?branch=develop)](https://travis-ci.org/OUXT-Polaris/rostate_machine)|

# How to use

## Example  

launch example state machine  
```
roslaunch rostate_machine example.launch  
```
if you succeed to launch, the message was appeared in the terminal  
```
[ INFO] [1556286488.574721992]: from : /example_event_client_node, succeed to transition, current state : emergency, event_name : motor_disconnected
[ INFO] [1556286488.674441095]: from : /example_event_client_node, failed to transition, current state : emergency, event_name : stop_motor
```
If you want to check whole state machine, you shold execute commands below
```
roscd rostate_machine/data
dot -T pdf example_state_machine.dot -o example_state_machine.pdf
evince example_state_machine.pdf
```
You can see pdf like below.  
![Example State Machine](img/example_state_machine.png "Example State Machine")

rostate_machine node and event_client communticate with state and event topic
![Example State Machine Node](img/rosgraph.png "Example State Machine Node")

State Topic (rostate_machine/State)
```
Header header
string current_state
string[] possible_transitions
string[] possible_transition_states
```

Event Topic (rostate_machine/Event)
```
Header header
string trigger_event_name
```

## Tutorial
### Make XML file in order to define state machine  

```
<state_machine>
    <init_state name="remote_operated"/>
    <state_machine_name name="example_state_machine"/>

    <callback tag="become_autonomos" when="on_entry" states="autonomous"/>
    <callback tag="stop_motor" when="on_exit" states="autonomous,remote_operated"/>
    <callback tag="check_motor" when="always" states="autonomous,remote_operated"/>

    <transition from = "remote_operated" to="autonomous" name="system_bringup"/>
    <transition from = "remote_operated" to="emergency" name="motor_disconnected"/>
    <transition from = "autonomous" to="remote_operated" name="manual_override"/>
    <transition from = "autonomous" to="emergency" name="motor_disconnected"/>
</state_machine>
```

transition element defines how to change states in the state machine.
rostate_machine node subscribe event topic and chenge state when the event name was matched with name attribute in the transition element in the XML.

callback element defines which callback function was called.
callback functions are managed with tag by using event_client.
states attribute in callback element defines the states in which calls callback.
states was defined by "," splited string.

when attribute defines when the callback function was called.
now, rostate_machine supports "always","on_exit","on_entry" values.
if you set when="always", callback function was called when the current_state topic was recieved and the current_state was matched with states value.
if you set when="on_exit" or when="on_entry", callback function was called when the current_state topic was recieved and the state was changed.
when="on_exit", the tagged callback functions was called when the state was changed from the setted states.
when="on_entry", the tagged callback functions was called when the state was changed to the setted states.

### Register Callback function to the EventClient class

add EventClient class to your ROS node and register callback by using EventClient::registerCallback function.

```
void EventClient::registerCallback(std::function<boost::optional<rostate_machine::Event>(void)> func,std::string tag)
```

### Make Launch files for your state machine
You can see how to make launch file for your statemachin in rostate_machine/launch/example.lannch

```
<include file="$(find rostate_machine)/launch/rostate_machine.launch">
    <arg name="xml_filepath" value="$(find rostate_machine)/data/example_state_machine.xml"/>
    <arg name="dot_filepath" value="$(find rostate_machine)/data/example_state_machine.dot"/>
    <arg name="state_machine_name" value="example_state_machine"/>
</include>

<node pkg="rostate_machine" type="example_event_client_node" name="example_event_client_node" output="log" respawn="false" respawn_delay="0">
    <param name="target_state_machine_namespace" value="/rostate_machine_node"/>
</node>
```

You have to pass xml_filepath,dot_filepath,state_machine_name args to the included launch file. (rostate_machine.launch)
You have to also set rosparam named (namespace)/target_state_machine_namespace.
Event client get state_machine_description from (value of (namespace)/target_state_machine_namespace )/xml_filepath

CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package rostate_machine
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.1 (2019-04-26)
------------------
* update README.md
* updat e README.md
* add doxygen comment
* add Doxygen Comment
* update .gitignroe
* add rosdoc.yaml
* remove unused topic
* fix function name
* remove unused headers
* remove unused header
* fix function getting active_tag
* remove StateChanged message
* enable publish trigger
* add TagInfo class
* add Callback Info Struct
* add callback xml parser
* add Callback Func Class
* add xml parser
* fix namespace
* add subscriber and publisher
* add sample event client
* add EventClient
* remove viewer
* add launch files
* add launch file
* add ROS executable
* add rostest
* add rostest to the depends
* add .rosinstall file
* add CI
* fix naming rules
* initial commit
* Contributors: Masaya Kataoka


src/rostate_machine.cpp
/**
 * @file rostate_machine.cpp
 * @author Masaya Kataoka (ms.kataoka@gmail.com)
 * @brief ROS Wrapper for the State Machine Library
 * @version 0.1
 * @date 2019-04-26
 * 
 * @copyright Copyright (c)  OUXT Polaris 2019
 * 
 */

#include <rostate_machine/rostate_machine.h>

/**
 * @brief Construct a new Rostate Machine:: Rostate Machine object
 * 
 * @param nh NodeHandle
 * @param pnh Private NodeHandle
 */
RostateMachine::RostateMachine(ros::NodeHandle nh,ros::NodeHandle pnh)
{
    nh_ = nh;
    pnh_ = pnh;
    //pnh_.param<std::string>("state_machine_name", state_machine_name_, "");
    pnh_.param<std::string>("dot_filepath", dot_filepath_, "");
    //pnh_.param<std::string>("xml_filepath", xml_filepath_, "");state_machine_name_
    pnh_.param<std::string>("description", description_, "");
}

/**
 * @brief Destroy the Rostate Machine:: Rostate Machine object
 * 
 */
RostateMachine::~RostateMachine()
{

}

/**
 * @brief Callback Function for Trigger Event Topic
 * 
 * @param event Event Message for the rostate_machine package.
 */
void RostateMachine::eventCallback(const ros::MessageEvent<rostate_machine::Event const>& event)
{
    rostate_machine::Event msg = *event.getMessage();
    StateInfo old_info = state_machine_ptr_->getStateInfo();
    bool result = state_machine_ptr_->tryTransition(msg.trigger_event_name);
    StateInfo info = state_machine_ptr_->getStateInfo();
    if(!result)
    {
        std::string publisher_name = event.getPublisherName();
        ROS_INFO_STREAM( "from : " << publisher_name << ", failed to transition, current state : "<< info.current_state << ", event_name : " << msg.trigger_event_name);
    }
    else
    {
        std::string publisher_name = event.getPublisherName();
        ROS_INFO_STREAM( "from : " << publisher_name << ", succeed to transition, current state : "<< info.current_state << ", event_name : " << msg.trigger_event_name);
    }
    return;
}

/**
 * @brief Run State Machine (Enable Transitions)
 * 
 */
void RostateMachine::run()
{
    state_machine_ptr_ = std::make_shared<StateMachine>(description_);
    state_machine_ptr_->drawStateMachine(dot_filepath_);
    nh_.param<double>(ros::this_node::getName()+"/publish_rate", publish_rate_, 10);
    //dot_string_pub_ = nh_.advertise<std_msgs::String>(ros::this_node::getName()+"/dot_string",1);
    current_state_pub_ = nh_.advertise<rostate_machine::State>(ros::this_node::getName()+"/"+state_machine_name_+"/current_state",1);

    boost::thread publish_thread(boost::bind(&RostateMachine::publishCurrentState, this));
    trigger_event_sub_ = nh_.subscribe(ros::this_node::getName()+"/"+state_machine_name_+"/trigger_event", 10, &RostateMachine::eventCallback,this);
    return;
}

/**
 * @brief Publish Current State Topic
 * 
 */
void RostateMachine::publishCurrentState()
{
    ros::Rate rate(publish_rate_);
    while(ros::ok())
    {
        //std_msgs::String dot_string_msg;
        //dot_string_msg.data = state_machine_ptr_->getDotString();
        //dot_string_pub_.publish(dot_string_msg);
        rostate_machine::State state_msg;
        StateInfo info = state_machine_ptr_->getStateInfo();
        state_msg.current_state = info.current_state;
        state_msg.possible_transitions = info.possibe_transitions;
        state_msg.possible_transition_states = info.possibe_transition_states;
        state_msg.header.stamp = ros::Time::now();
        current_state_pub_.publish(state_msg);
        rate.sleep();
    }
    return;
}